src/Map.php

```diff
@@ -210,34 +283,52 @@ private function keysAreEqual($a, $b): bool
      * Attempts to look up a key in the table.
      *
      * @param $key
+     * @param $hash The internal hash computed for the key.
+     *
+     * @psalm-param TKey $key
+     * @psalm-param-out string|int $hash
      *
-     * @return Pair|null
+     * @return PairRef<TKey, TValue>|null
      *
-     * @psalm-return Pair<TKey, TValue>|null
+     * @psalm-return PairRef<TKey, TValue>|null
      */
-    private function lookupKey($key)
+    private function lookupKey($key, &$hash = null)
     {
-        foreach ($this->pairs as $pair) {
-            if ($this->keysAreEqual($pair->key, $key)) {
-                return $pair;
+        $hash = $this->getHash($key);
+        if (!array_key_exists($hash, $this->table)) {
+            return null;
+        }
+
+        $hashMatch = $this->table[$hash];
+        if ($hashMatch instanceof PairRef) {
```

Comment:
is there any instance where it wont be an array?

Reply:
In order to save on memory, we start out by associating each hash key with just a `PairRef`. We only convert the value associated with a hash key to `list<PairRef>` once we encounter a collision (two different values that hash to the same thing); at that point we'll put the first value and the new value into a new array, which will be size 2. This saves on memory because, in PHP, initializing an array with one thing in it takes up a lot of extra memory compared to just the one thing on its own, and since we expect collisions to be rare, we don't often have to change up from the single `PairRef` to the `list<PairRef>`. This also ends up being faster, I believe because we're avoiding a bunch of allocations for the arrays.

Reply:
ok i see i was looking at an old commit and you change it